src/builtins/set.cpp

```diff
@@ -620,45 +623,50 @@ static int builtin_set_show(const wchar_t *cmd, const set_cmd_opts_t &opts, int
 static int builtin_set_erase(const wchar_t *cmd, set_cmd_opts_t &opts, int argc,
                              const wchar_t **argv, parser_t &parser, io_streams_t &streams) {
     int ret = STATUS_CMD_OK;
-    env_mode_flags_t scope = compute_scope(opts);
-    for (int i = 0; i < argc; i++) {
-        auto split = split_var_and_indexes(argv[i], scope, parser.vars(), streams);
-        if (!split) {
-            builtin_print_error_trailer(parser, streams.err, cmd);
-            return STATUS_CMD_ERROR;
-        }
+    env_mode_flags_t scopes = compute_scope(opts);
+    // `set -e` is allowed to be called with multiple scopes.
+    for (int bit = 0; 1<<bit <= ENV_USER; ++bit) {
+        int scope = scopes & 1<<bit;
+        if (scope == 0 || (scope == ENV_USER && scopes != ENV_USER)) continue;
+        for (int i = 0; i < argc; i++) {
+            auto split = split_var_and_indexes(argv[i], scope, parser.vars(), streams);
```

Comment:
I'm assuming with indexes this works like:

```fish
set -g foo 1 2 3
set -l foo a b c

set -egl foo[2]
```

makes the global foo "1 3" and the local foo "a c"?

Reply:
That is indeed how it works, and it's probably a good test case.

Variable events are also fired for each scope, which is probably okay for now.

(tbh I'm still of the opinion local variables shouldn't fire events - you can't read them in an event handler anyway - but last time that turned out to be unpopular)

Reply:
Yes, what I had in mind the entire time was that `set -elg foo` should behave exactly as if `set -el foo; set -eg foo` (except for the return codes, which are already somewhat non-sensical and need to be fixed later), side effects and all. We can always change that behavior later if we have a good reason to, but I wanted to minimize the scope of the changes for this PR to make this an easier yes/no decision.

> (tbh I'm still of the opinion local variables shouldn't fire events - you can't read them in an event handler anyway - but last time that turned out to be unpopular)

I must have missed that discussion. Do you happen to remember what issue this was in?

Reply:
> I must have missed that discussion. Do you happen to remember what issue this was in?

#8384, which asked about `for`-loop variables.

In that case, we chose to keep it because someone could try to do something with `--no-scope-shadowing`.